### PR TITLE
[win32] Fix scaling of imagelist with gaps

### PR DESCRIPTION
**Pull request by akoch-yatta**
When an ImageList is scaled (e g. when a ToolBar is scaled due to a zoom changed) null images in the images array (e.g. when a ToolItem is disposed) were not addressed properly. This led to scenarios like this:

Starting on 200%
image

Without PR on 100% -> The behaviour depends on the index where the images are missing and when the ImageList in scales. In some scenarios, wrong images will be shown, in this scenario, images are missing
image

With PR on 100%
image

If a null image is found when scaling an ImageList a valid placeholder image is added to the ImageList.

## How to test

```
import org.eclipse.swt.*;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.widgets.*;

public class ToolbarTest {

	public static void main (String [] args) {
		Display display = new Display ();
		Shell shell = new Shell(display);
		shell.setLayout(new RowLayout());
		shell.setText("ToolbarTest");

		Image errorImage = display.getSystemImage(SWT.ICON_ERROR);
		Image infoImage = display.getSystemImage(SWT.ICON_INFORMATION);

		ToolBar imageToolBar = new ToolBar (shell, SWT.NONE);
		ToolItem item = new ToolItem (imageToolBar, SWT.PUSH);
		item.setImage (errorImage);
		item = new ToolItem (imageToolBar, SWT.PUSH);
		item.setImage (errorImage);
		item = new ToolItem (imageToolBar, SWT.RADIO);
		item.setImage (errorImage);
		item = new ToolItem (imageToolBar, SWT.RADIO);
		item.setImage (errorImage);
		ToolItem item2 = new ToolItem (imageToolBar, SWT.RADIO);
		item2.setImage (infoImage);
		ToolItem item3 = new ToolItem (imageToolBar, SWT.RADIO);
		item3.setImage (infoImage);
		item = new ToolItem (imageToolBar, SWT.RADIO);
		item.setImage (errorImage);
		item = new ToolItem (imageToolBar, SWT.RADIO);
		item.setImage (errorImage);
		item = new ToolItem (imageToolBar, SWT.SEPARATOR);
		item.setToolTipText("SWT.SEPARATOR");

		item2.dispose();
		item3.dispose();

		shell.pack();
		shell.open ();
		while (!shell.isDisposed ()) {
			if (!display.readAndDispatch ()) display.sleep ();
		}
		display.dispose ();
	}
}
```

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 1]
```diff
@@ -389,10 +396,10 @@ private void setForAllHandles(int index, Image image, int count) {
 	zoomToHandle.forEach((zoom, handle) -> set(index, image, count, handle, zoom));
 }
 
-void set (int index, Image image, int count, long handle, int zoom) {
+void set (int index, Image image, int count, long listHandle, int zoom) {
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:399]: Is this parameter renaming necessary/reasonable?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 2]
```diff
@@ -334,16 +334,23 @@ public long getHandle(int targetZoom) {
...
-				count++;
+				set(i, image, i, newImageListHandle, targetZoom);
+			} else {
+				long hDC = OS.GetDC (0);
+				if (hDC == 0) SWT.error(SWT.ERROR_NO_HANDLES);
+				long placeholderBitmapHandle = OS.CreateCompatibleBitmap(hDC, scaledWidth, scaledHeight);
+				if (placeholderBitmapHandle == 0) SWT.error(SWT.ERROR_NO_HANDLES);
+				OS.ImageList_Add(newImageListHandle, placeholderBitmapHandle, placeholderBitmapHandle);
+				OS.DeleteObject(placeholderBitmapHandle);
+				OS.ReleaseDC(0, hDC);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:350]: Would be good to extract this into a method with a self-describing name.

**Review by HeikoKlare — COMMENTED** [review]
The change looks sound to me. I only have minor comments.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 1 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:399]: It is shadowing the class attribute handle, so I would strongly argument for changing it. That is an easy source for a bug

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 2 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: I extracted it into a separate method

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 1 above)
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:403]: I see. I actually didn't want to challenge the change itself but understand how it is related to the PR's goal. If I understand correctly, it is a refactoring that could be done independently. But it's also fine for me to merge that simple change with this PR.

**Review by HeikoKlare — APPROVED** [review]
The change looks good anf works as expected. I tested with both the provided snippet (for checking the enhanced functionality) and with the SDK (to identify potential regressions).

**Pull request merged**